File: modules/routeplanner/src/main/java/com/sygic/modules/routeplanner/di/module/ViewModelModule.java
package com.sygic.modules.routeplanner.di.module;

import com.sygic.modules.common.di.util.ViewModelKey;
import com.sygic.modules.common.di.util.ViewModelModuleBase;
import com.sygic.modules.routeplanner.viewmodel.RoutePlannerFragmentViewModel;
import com.sygic.modules.routeplanner.viewmodel.RoutePlannerFragmentViewModelFactory;
import com.sygic.tools.viewmodel.ViewModelCreatorFactory;
import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module(
        includes = {
                ViewModelModuleBase.class
        }
)
public abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(RoutePlannerFragmentViewModel.class)
    abstract ViewModelCreatorFactory putRoutePlannerFragmentViewModelFactory(RoutePlannerFragmentViewModelFactory factory);
}

File: modules/common/src/main/java/com/sygic/modules/common/di/ViewModelModule.kt
package com.sygic.modules.common.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import dagger.Binds
import dagger.MapKey
import dagger.Module
import javax.inject.Inject
import javax.inject.Provider
import kotlin.reflect.KClass

interface ViewModelCreatorFactory {
    fun create(vararg assistedValues: Any?): ViewModel
}

@Suppress("UNCHECKED_CAST")
class ViewModelFactory @Inject constructor(private val viewModels: MutableMap<Class<out ViewModel>, Provider<ViewModelCreatorFactory>>) :
    ViewModelProvider.Factory {

    private var assistedValues: Array<out Any?> = emptyArray()

    override fun <T : ViewModel> create(modelClass: Class<T>): T = viewModels[modelClass]?.get()?.create(*assistedValues) as T

    fun with(vararg assistedValues: Any?): ViewModelFactory {
        this.assistedValues = assistedValues
        return this
    }
}

@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
@MapKey
annotation class ViewModelKey(val value: KClass<out ViewModel>)

@Module
abstract class ViewModelModuleBase {

    @Binds
    protected abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory

}